File: codes/HellDream/47.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> lists = new ArrayList<>();
        List<Integer> tmp = new ArrayList<>();
        permutation(lists, tmp, new boolean[nums.length], nums);
        return lists;
    }
        public void permutation(List<List<Integer>> lists,List<Integer> tmp, boolean[] used, int[] nums){
        if(tmp.size()==nums.length){
            lists.add(tmp);
            return;
        }
        for(int i=0;i<nums.length;i++){
            if(used[i]||(i>0&&nums[i]==nums[i-1]&&!used[i-1])){
                continue;
            }else{
                List<Integer> list = new ArrayList<>(tmp);
                list.add(nums[i]);
                used[i] = true;
                permutation(lists, list, used, nums);
                used[i] = false;
            }
        }
    }
}

File: codes/Garnetwzy/47.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ret;
        vector<bool> visit(nums.size(), false);
        vector<int> cur = {};
        dfs(ret, cur, nums, visit);
        return ret;
    }
    
    void dfs(vector<vector<int>>& ret, vector<int>& cur, vector<int>& nums,  vector<bool> visit) {
        if(cur.size() == nums.size()) {
            ret.push_back(cur);
            return;
        }
        
        int i = 0;
        int j;
        while(i < nums.size()) {
            if(visit[i]) {
                i++;
                continue;
            }
            visit[i] = true;
            cur.push_back(nums[i]);
            dfs(ret, cur, nums, visit);
            visit[i] = false;
            cur.pop_back();
            j = i+1;
            while(j < nums.size() && nums[j] == nums[j-1]) {
                j++;
            }
            i = j;
        }
    }
};